### ISSUE TITLE
Display P3 color profile not respected with HEIC file

### ISSUE THREAD

**Issue by Andreas0602** [jurplel/qView#606]
**Environment:**
- OS: macOS 13.5 Ventura
- qView version: nightly 2023-08-08.1

**Information:**

[Demo.zip](https://github.com/jurplel/qView/files/12289996/Demo.zip)

When showing the HEIC image in the attached ZIP file, qView will not take the image's color profile "Display P3" into account. As a result, colors will be shown incorrectly.

When showing the same image in JPG format (from the attached ZIP file), the problem does not occur.

You can easily see the problem when you open both files in qView, look at the red jacket of the man on left side of the picture, and quickly toggle between both windows by means of the keyboard shortcut for moving the focus to the next window (`command` `<` on a German keyboard, for other languages see [here](https://apple.stackexchange.com/questions/193937/shortcut-for-toggling-between-different-windows-of-same-app)).

**Comment by jurplel** [comment]
@jdpurcell Correct me if I'm wrong but I think the current implementation only checks for the color space of the monitor.

**Comment by jurplel** [comment]
oh, interesting. Maybe the built-in Qt HEIC image plugin does not support color profiles. The kimageformats one definitely does as can be seen here: https://invent.kde.org/frameworks/kimageformats/-/merge_requests/16/diffs#8ff64707ffb67922cb5aa27b7e57623fdc0d8e87

Sure looks like the built-in Qt one doesn't: https://github.com/qt/qtimageformats/blob/dev/src/plugins/imageformats/macheif/qmacheifhandler.cpp

Maybe this can be resolved by changing the image format plugin to the one from kimageformats, then.

**Comment by jurplel** [comment]
Oh that is really strange indeed. XnViewMP should have nothing in common with qView codebase wise.

**Comment by jdpurcell** [comment]
qView's implementation does consider both the image's and the monitor's color space. However for this image, using Qt's HEIC support, I can see qView isn't finding a color space in the image and assuming it's sRGB. Interestingly that works out pretty well in this case, at least judging by the image looks compared to the Preview app, as qView looks very similar although the red clothing isn't as vibrant. If I make it assume the image Display P3 it actually looks even more off, with everything looking oversaturated. I suspect what's happening is that Qt's HEIC decoder is factoring in the color space, but instead of decoding it as-is and reporting the color space, it converts the decoded image to sRBG automatically and doesn't expose anything in `QImage::colorSpace()`.

So as @jurplel suggested, if kimageformat's HEIC decoder works better in this regard, it could help. I don't know how to get it working though so I'm unable to do any testing.

**Comment by jurplel** [comment]
Very interesting--thank you for the detailed writeup!

For this upcoming version, since I actually want to ship a version before the heat death of the universe, I am going to skip trying to get kimageformats HEIF plugin building on macOS--I had some issues with it so I cut it for now. Not sure if it would help anyway based on your description.

**Comment by jdpurcell** [comment]
@jurplel Based on the limited testing I could figure out how to do, i.e. a special build with some additional [debug output](https://github.com/jdpurcell/qView/commit/a76c37c6d511b4ec0f9941101b2e0ac57d5ca2a7), I am seeing that on macOS it's not detecting a color space from the image, whereas on Windows it detected Display P3 from the image. So assuming the difference is due to the Windows version having the kimageformats-based plugin, it may indeed be worthwhile to get the kimageformats HEIF plugin building on macOS if at all practical.

**Comment by jurplel** [comment]
Yeah, I should work on that at some point.

**Comment by jdpurcell** [comment]
Good news! I got the kimageformats HEIF plugin building. I think it was resolved after a package update from bumping `vcpkgGitCommitId`, but I made a lot of other changes recently in my kimageformats-binaries fork so I'm not sure.

As for the qView side, I just needed to [avoid including](https://github.com/jdpurcell/qView/commit/64c545944c571b991c7ffb0391672462af9715bc) the HEIF plugin from qtimageformats. I tested with the demo images provided in the original bug report, and it fixes the color difference on the red jacket.
